<a href="https://colab.research.google.com/github/nikshargithub/ML_PROJECTS/blob/main/2024_01_14_NIKHIL_SHARMA_PROJECT_78.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Instructions

---

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concepts covered in the Logistic Regression Lessons





---

#### Getting Started:

1. Follow the next 3 steps to create a copy of this colab file and start working on the project.
     
2. Create a duplicate copy of the Colab file as described below.

  - Click on the **File menu**. A new drop-down list will appear.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/0_file_menu.png' width=500>

  - Click on the **Save a copy in Drive** option. A duplicate copy will get created. It will open up in the new tab on your web browser.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/lesson-0/1_create_colab_duplicate_copy.png' width=500>

3. After creating the duplicate copy of the notebook, please rename it in the **YYYY-MM-DD_StudentName_Project78** format.

4. Now, write your code in the prescribed code cells.


---

### Problem Statement

Nowadays everyone possesses one or more email accounts for digital communication. You may have observed that your mailbox classifies your relevant emails as primary emails and other irrelevant emails as spam.

**Spam** emails are junk emails or unrequired emails. They may consist of advertisements, updated, and unwanted messages which are sent to the receivers without their permission. Hence, spam detection is one of the important problem statement in email and messaging servicing.

In this project, you will create a model to classify the emails as **spam** or **not spam** using Logistic Regression.


---

### List of Activities

**Activity 1:** Analysing the Dataset

**Activity 2:** Train-Test Split

**Activity 3:** Normalisation of the Features

**Activity 4:** Logistic Regression - Model Training

**Activity 5:** Logistic Regression - Model Prediction and Evaluation










---


#### Activity 1:  Analysing the Dataset

Create a Pandas DataFrame for **Spambase** dataset using the below link. This dataset consists of the following 58 columns. Most of the columns represent frequencies of a particular word or character in the email:


|Column #|Attribute|Description|
|-|-|-|
|0 - 47|word_freq_WORD|The first 48 columns are the percentage|
|||of the frequencies of the particular word one column each|
||||
|48 - 53|word_freq_CHAR|The next 6 columns are the percentage of the frequencies of the|
|||particular special character like semi-colon (;), exclamation (!) one column each|
||||
|54|capital_run_length_average|average length of uninterrupted sequences of capital letters|
||||
|55|capital_run_length_longest|length of longest uninterrupted sequence of capital letters|
||||
|56|capital_run_length_total|sum of the length of uninterrupted sequences of capital letters|
||||
|57|email_type|denotes whether the email is spam (1) or not (0)|



 -  **Dataset Link:** https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/spambase.csv
   
- **Source:** UCI Library https://archive.ics.uci.edu/ml/datasets/Spambase






Print the first five rows of the dataset. Check for null values and treat them accordingly (if any).

In [ ]:
# Import modules
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import math
# Load the dataset
df = pd.read_csv('https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/spambase.csv')
# Print the first five rows of the DataFrame
df.head()

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.41  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.42  0.43  0.778   0.44   0.45  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

Rename the last column of the DataFrame as target.

In [ ]:
# Rename the last column as 'target'
df.rename(columns = {df.columns[-1]:'target'},inplace = True)

Print the information of the DataFrame to verify the above update.

In [ ]:
# Print the dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4600 non-null   float64
 1   0.64    4600 non-null   float64
 2   0.64.1  4600 non-null   float64
 3   0.1     4600 non-null   float64
 4   0.32    4600 non-null   float64
 5   0.2     4600 non-null   float64
 6   0.3     4600 non-null   float64
 7   0.4     4600 non-null   float64
 8   0.5     4600 non-null   float64
 9   0.6     4600 non-null   float64
 10  0.7     4600 non-null   float64
 11  0.64.2  4600 non-null   float64
 12  0.8     4600 non-null   float64
 13  0.9     4600 non-null   float64
 14  0.10    4600 non-null   float64
 15  0.32.1  4600 non-null   float64
 16  0.11    4600 non-null   float64
 17  1.29    4600 non-null   float64
 18  1.93    4600 non-null   float64
 19  0.12    4600 non-null   float64
 20  0.96    4600 non-null   float64
 21  0.13    4600 non-null   float64
 22  

In [ ]:
df.isnull().sum()

0         0
0.64      0
0.64.1    0
0.1       0
0.32      0
0.2       0
0.3       0
0.4       0
0.5       0
0.6       0
0.7       0
0.64.2    0
0.8       0
0.9       0
0.10      0
0.32.1    0
0.11      0
1.29      0
1.93      0
0.12      0
0.96      0
0.13      0
0.14      0
0.15      0
0.16      0
0.17      0
0.18      0
0.19      0
0.20      0
0.21      0
0.22      0
0.23      0
0.24      0
0.25      0
0.26      0
0.27      0
0.28      0
0.29      0
0.30      0
0.31      0
0.33      0
0.34      0
0.35      0
0.36      0
0.37      0
0.38      0
0.39      0
0.40      0
0.41      0
0.42      0
0.43      0
0.778     0
0.44      0
0.45      0
3.756     0
61        0
278       0
target    0
dtype: int64

**Q:** Are there any missing values?

**A:**No

---


#### Activity 2: Train-Test Split

You have to determine the effect of all the features on the 'target' variable. Thus, every column other than the `target` is the feature variable and `target` column is the target variable.

**Steps:**

1. Create a list of all the features.

2. Split the "DataFrame" into features and target arrays using the features list.

3. Split the dataset into a training set and test set such that the training set contains 70% of the instances and the remaining instances will become the test set.

4. Reshape the target variable arrays into two-dimensional arrays by using `reshape(-1, 1)` function of the `numpy` module.


In [ ]:
# Split the DataFrame into the train and test sets.

# Import the module
from sklearn.model_selection import train_test_split

# Create a features list
features = list(df.columns)
features.remove('target')
print(features)
x = df[features]
y = df['target']
# Split the DataFrame into the train and test sets such that test set has 30% of the values.
X_train,X_test,y_train,y_test = train_test_split(x, y, test_size = 0.30, random_state = 42)
# Reshape target arrays to 2-dimensional array.
y_train_reshaped = np.array(y_train).reshape(-1,1)
y_test_reshaped = np.array(y_test).reshape(-1,1)
print(X_train.shape)
print(X_test.shape)
print(y_train_reshaped.shape)
print(y_test_reshaped.shape)

['0', '0.64', '0.64.1', '0.1', '0.32', '0.2', '0.3', '0.4', '0.5', '0.6', '0.7', '0.64.2', '0.8', '0.9', '0.10', '0.32.1', '0.11', '1.29', '1.93', '0.12', '0.96', '0.13', '0.14', '0.15', '0.16', '0.17', '0.18', '0.19', '0.20', '0.21', '0.22', '0.23', '0.24', '0.25', '0.26', '0.27', '0.28', '0.29', '0.30', '0.31', '0.33', '0.34', '0.35', '0.36', '0.37', '0.38', '0.39', '0.40', '0.41', '0.42', '0.43', '0.778', '0.44', '0.45', '3.756', '61', '278']
(3220, 57)
(1380, 57)
(3220, 1)
(1380, 1)


----

#### Activity 3: Normalisation of the Features

Get a descriptive analysis of the feature set and decide whether any normalisation is needed.

Describe the features for training data.

In [ ]:
# Get the descriptive statistics for 'X_train'.
X_train.describe()

0         0.64       0.64.1          0.1         0.32  \
count  3220.000000  3220.000000  3220.000000  3220.000000  3220.000000   
mean      0.105590     0.194068     0.284668     0.061326     0.315283   
std       0.306026     1.200963     0.516486     1.333192     0.691309   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.420000     0.000000     0.380000   
max       4.540000    14.280000     5.100000    42.810000    10.000000   

               0.2          0.3          0.4          0.5          0.6  ...  \
count  3220.000000  3220.000000  3220.000000  3220.000000  3220.000000  ...   
mean      0.098438     0.106609     0.103963     0.088748     0.234516  ...   
std       0.280005     0.373671     0.389894     0.255005     0.642719  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.150000  ...   
max       5.880000     7.270000    11.110000     3.330000    18.180000  ...   

              0.40         0.41         0.42         0.43        0.778  \
count  3220.000000  3220.000000  3220.000000  3220.000000  3220.000000   
mean      0.033547     0.036403     0.142218     0.016734     0.268700   
std       0.274906     0.231370     0.288173     0.117969     0.892478   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.067000     0.000000     0.000000   
75%       0.000000     0.000000     0.192000     0.000000     0.310000   
max       8.330000     4.187000     9.752000     4.081000    32.478000   

              0.44         0.45        3.756           61           278  
count  3220.000000  3220.000000  3220.000000  3220.000000   3220.000000  
mean      0.074947     0.044727     4.968248    52.542857    276.737888  
std       0.263572     0.368815    29.236502   216.614341    559.771635  
min       0.000000     0.000000     1.000000     1.000000      1.000000  
25%       0.000000     0.000000     1.583000     6.000000     34.000000  
50%       0.000000     0.000000     2.252500    14.000000     92.000000  
75%       0.048000     0.000000     3.628250    41.000000    260.000000  
max       6.003000    13.129000  1021.500000  9989.000000  10062.000000  

[8 rows x 57 columns]

Describe the features for the testing data.

In [ ]:
# Get the descriptive statistics for 'X_test'.
X_test.describe()

0         0.64       0.64.1          0.1         0.32  \
count  1380.000000  1380.000000  1380.000000  1380.000000  1380.000000   
mean      0.102210     0.256913     0.271036     0.075036     0.305080   
std       0.303988     1.478609     0.474234     1.530927     0.626911   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.420000     0.000000     0.390000   
max       4.000000    14.280000     3.700000    42.730000     6.250000   

               0.2          0.3          0.4          0.5          0.6  ...  \
count  1380.000000  1380.000000  1380.000000  1380.000000  1380.000000  ...   
mean      0.090051     0.132022     0.108478     0.093210     0.251014  ...   
std       0.258923     0.429796     0.426271     0.327307     0.649769  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%       0.000000     0.000000     0.000000     0.000000     0.170000  ...   
max       3.570000     7.270000     5.880000     5.260000    11.110000  ...   

              0.40         0.41         0.42         0.43        0.778  \
count  1380.000000  1380.000000  1380.000000  1380.000000  1380.000000   
mean      0.027978     0.043669     0.131694     0.017551     0.269567   
std       0.309697     0.269704     0.223330     0.086214     0.599871   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.059500     0.000000     0.000000   
75%       0.000000     0.000000     0.177000     0.000000     0.333250   
max      10.000000     4.385000     2.941000     1.449000     8.333000   

              0.44         0.45        3.756           61           278  
count  1380.000000  1380.000000  1380.000000  1380.000000   1380.000000  
mean      0.077880     0.043129     5.713512    51.302899    298.579710  
std       0.198743     0.545316    36.913490   131.034024    703.306895  
min       0.000000     0.000000     1.000000     1.000000      1.000000  
25%       0.000000     0.000000     1.600000     6.000000     37.000000  
50%       0.000000     0.000000     2.310500    16.000000    101.500000  
75%       0.061000     0.000000     3.852500    47.000000    287.250000  
max       3.125000    19.829000  1102.500000  2204.000000  15841.000000  

[8 rows x 57 columns]

**Q:** Does the data needs normalisation? Why?

**A:**Yes, maximum values varies a lot


If the answer to the above question is **yes**, Normalise the data by calculating their $Z$-scores (or standard scaler) in the following code sections.

Define the Standard Normalisation function.

In [ ]:
# Define the 'standard_scalar()' function for calculating Z-scores
def standard_scalar(ser):
  z = (ser-ser.mean())/ser.std()
  return z

**Hint**
$Z$-score for each value can be calculated by the following expression:

$$Z = \frac{X - \mu}{\sigma}$$

Where,


- $X$ is an observation

- $\mu$ is the population mean

- $\sigma$ is the population standard deviation


Apply the normalisation function to the features of the training data.

In [ ]:
# Apply the 'standard_scalar()' on X_train using apply() function and get the descriptive statistics of the normalised X_train
norm_X_train = X_train.apply(standard_scalar, axis=0)
norm_X_train.describe()

0          0.64        0.64.1           0.1          0.32  \
count  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03   
mean   2.758318e-18  2.206654e-17 -1.544658e-17  7.723291e-18 -1.323993e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.450362e-01 -1.615939e-01 -5.511625e-01 -4.599943e-02 -4.560661e-01   
25%   -3.450362e-01 -1.615939e-01 -5.511625e-01 -4.599943e-02 -4.560661e-01   
50%   -3.450362e-01 -1.615939e-01 -5.511625e-01 -4.599943e-02 -4.560661e-01   
75%   -3.450362e-01 -1.615939e-01  2.620251e-01 -4.599943e-02  9.361571e-02   
max    1.449030e+01  1.172886e+01  9.323259e+00  3.206490e+01  1.400924e+01   

                0.2           0.3           0.4           0.5           0.6  \
count  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03   
mean  -6.178632e-17 -1.710157e-17 -7.171627e-18  2.978984e-17  2.206654e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.515572e-01 -2.853011e-01 -2.666435e-01 -3.480258e-01 -3.648803e-01   
25%   -3.515572e-01 -2.853011e-01 -2.666435e-01 -3.480258e-01 -3.648803e-01   
50%   -3.515572e-01 -2.853011e-01 -2.666435e-01 -3.480258e-01 -3.648803e-01   
75%   -3.515572e-01 -2.853011e-01 -2.666435e-01 -3.480258e-01 -1.314968e-01   
max    2.064804e+01  1.917033e+01  2.822827e+01  1.271052e+01  2.792119e+01   

       ...          0.40          0.41          0.42          0.43  \
count  ...  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03   
mean   ... -1.710157e-17  1.323993e-17 -7.061294e-17  6.619963e-17   
std    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min    ... -1.220294e-01 -1.573388e-01 -4.935153e-01 -1.418549e-01   
25%    ... -1.220294e-01 -1.573388e-01 -4.935153e-01 -1.418549e-01   
50%    ... -1.220294e-01 -1.573388e-01 -2.610159e-01 -1.418549e-01   
75%    ... -1.220294e-01 -1.573388e-01  1.727515e-01 -1.418549e-01   
max    ...  3.017926e+01  1.793925e+01  3.334728e+01  3.445201e+01   

              0.778          0.44          0.45         3.756            61  \
count  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03  3.220000e+03   
mean   2.978984e-17 -2.427320e-17  1.654991e-17  2.316987e-17 -1.323993e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.010719e-01 -2.843524e-01 -1.212732e-01 -1.357292e-01 -2.379476e-01   
25%   -3.010719e-01 -2.843524e-01 -1.212732e-01 -1.157884e-01 -2.148651e-01   
50%   -3.010719e-01 -2.843524e-01 -1.212732e-01 -9.288895e-02 -1.779331e-01   
75%    4.627566e-02 -1.022387e-01 -1.212732e-01 -4.583305e-02 -5.328759e-02   
max    3.608975e+01  2.249125e+01  3.547654e+01  3.476927e+01  4.587165e+01   

                278  
count  3.220000e+03  
mean  -4.854640e-17  
std    1.000000e+00  
min   -4.925900e-01  
25%   -4.336373e-01  
50%   -3.300237e-01  
75%   -2.990128e-02  
max    1.748081e+01  

[8 rows x 57 columns]

Apply the normalisation function to the features of the test data.

In [ ]:
# Apply the 'standard_scalar()' on X_test and get the descriptive statistics of the normalised X_test
norm_X_test = X_test.apply(standard_scalar, axis=0)
norm_X_test.describe()

0          0.64        0.64.1           0.1          0.32  \
count  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03   
mean  -2.767512e-17 -2.059544e-17  3.089316e-17  1.287215e-18  2.059544e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.362312e-01 -1.737532e-01 -5.715240e-01 -4.901358e-02 -4.866398e-01   
25%   -3.362312e-01 -1.737532e-01 -5.715240e-01 -4.901358e-02 -4.866398e-01   
50%   -3.362312e-01 -1.737532e-01 -5.715240e-01 -4.901358e-02 -4.866398e-01   
75%   -3.362312e-01 -1.737532e-01  3.141144e-01 -4.901358e-02  1.354583e-01   
max    1.282220e+01  9.483975e+00  7.230529e+00  2.786217e+01  9.482882e+00   

                0.2           0.3           0.4           0.5           0.6  \
count  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03   
mean   2.574430e-18  2.445709e-17  6.178632e-17 -1.029772e-17 -1.287215e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -3.477892e-01 -3.071728e-01 -2.544822e-01 -2.847794e-01 -3.863134e-01   
25%   -3.477892e-01 -3.071728e-01 -2.544822e-01 -2.847794e-01 -3.863134e-01   
50%   -3.477892e-01 -3.071728e-01 -2.544822e-01 -2.847794e-01 -3.863134e-01   
75%   -3.477892e-01 -3.071728e-01 -2.544822e-01 -2.847794e-01 -1.246820e-01   
max    1.344008e+01  1.660781e+01  1.353957e+01  1.578578e+01  1.671207e+01   

       ...          0.40          0.41          0.42          0.43  \
count  ...  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03   
mean   ...  2.574430e-18  1.029772e-17  4.119088e-17  1.802101e-17   
std    ...  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min    ... -9.034086e-02 -1.619142e-01 -5.896831e-01 -2.035800e-01   
25%    ... -9.034086e-02 -1.619142e-01 -5.896831e-01 -2.035800e-01   
50%    ... -9.034086e-02 -1.619142e-01 -3.232618e-01 -2.035800e-01   
75%    ... -9.034086e-02 -1.619142e-01  2.028644e-01 -2.035800e-01   
max    ...  3.219932e+01  1.609668e+01  1.257914e+01  1.660343e+01   

              0.778          0.44          0.45         3.756            61  \
count  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03  1.380000e+03   
mean   3.604202e-17 -5.406303e-17 -2.574430e-18 -2.574430e-18  2.059544e-17   
std    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
min   -4.493757e-01 -3.918651e-01 -7.908990e-02 -1.276908e-01 -3.838919e-01   
25%   -4.493757e-01 -3.918651e-01 -7.908990e-02 -1.114365e-01 -3.457339e-01   
50%   -4.493757e-01 -3.918651e-01 -7.908990e-02 -9.218883e-02 -2.694178e-01   
75%    1.061605e-01 -8.493601e-02 -7.908990e-02 -5.041549e-02 -3.283802e-02   
max    1.344195e+01  1.533196e+01  3.628331e+01  2.971235e+01  1.642854e+01   

                278  
count  1.380000e+03  
mean  -2.188266e-17  
std    1.000000e+00  
min   -4.231150e-01  
25%   -3.719283e-01  
50%   -2.802187e-01  
75%   -1.610920e-02  
max    2.209906e+01  

[8 rows x 57 columns]

---

#### Activity 4: Logistic Regression - Model Training

Implement Logistic Regression Classification using `sklearn` module to estimate the values of $\beta$ coefficients in the following way:

1. Deploy the model by importing the `LogisticRegression` class and create an object of this class.
2. Call the `fit()` function on the Logistic Regression object and print score using the `score()` function.
3. Print the $\beta$ coefficient values.

In [ ]:
# Deploy the 'LogisticRegression' model using the 'fit()' function.
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train_reshaped)
print(log_reg.score(X_train, y_train_reshaped))

0.9301242236024845


/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Get the beta coefficients for the features using the model object trained in the above code.

In [ ]:
# Print the beta coefficient values
print(log_reg.coef_)

[[-8.71927849e-02 -1.57863799e-01  9.56956751e-02  2.59690879e-01
   1.06041852e+00  5.62626970e-01  1.37159388e+00  1.01838948e+00
   4.18131690e-01  1.48932481e-01  1.14416298e-01 -4.60960455e-02
   1.07573103e-01 -2.88670325e-04  2.68495359e-01  1.53070351e+00
   9.20585036e-01  2.94414161e-01 -7.44556412e-03  5.13025743e-01
   2.51589700e-01  3.11022829e-01  1.21069881e+00  4.95940496e-01
  -2.77785408e+00 -1.36979201e+00 -3.54533872e+00 -3.49834507e-01
  -6.86820403e-01 -4.82733263e-01 -4.21119026e-01 -2.56521252e-01
  -6.75613322e-01 -2.60222455e-01 -6.39849898e-01 -1.58702975e-01
  -4.74629382e-01 -2.10821978e-01 -4.69808086e-01 -2.39157458e-01
  -3.35014407e-01 -1.05076587e+00 -2.93844147e-01 -7.14448013e-01
  -9.85717822e-01 -1.42408651e+00 -8.76922657e-02 -3.27881874e-01
  -4.42021081e-01 -1.89554997e-01 -1.13774461e-01  3.22258144e-01
   8.00334937e-01  2.72780239e-01 -1.38004378e-02  7.94871440e-03
   7.15492624e-04]]


----

#### Activity 5: Logistic Regression - Model Prediction and Evaluation

Predict the values for both training and test sets by calling the `predict()` function on the Logistic Regression object.

In [ ]:
# Make predictions on the test dataset by using the 'predict()' function.
y_test_pred = log_reg.predict(X_test)
y_test_pred

array([0, 0, 0, ..., 0, 1, 0])

Also, display the confusion matrix.

In [ ]:
# Display the results of confusion_matrix
from sklearn.metrics import confusion_matrix, classification_report
confusion_matrix(y_test, y_test_pred)

array([[745,  58],
       [ 78, 499]])

**Q:** What is the positive outcome out of both the labels?

**A:**Positive Outcomes = 499 + 745 = 1244

**Q:** Write the count of True Positives and True Negatives?

**A:** TP:499   TN:78


Print the classification report values to evaluate the accuracy of your model.

In [ ]:
# Display the results of classification_report
print(classification_report(y_test, y_test_pred))

              precision    recall  f1-score   support

           0       0.91      0.93      0.92       803
           1       0.90      0.86      0.88       577

    accuracy                           0.90      1380
   macro avg       0.90      0.90      0.90      1380
weighted avg       0.90      0.90      0.90      1380



**Q** Write the f1-score of both labels?

**A:**0:0.92 1:0.88

-----

### Submitting the Project

1. After finishing the project, click on the **Share** button on the top right corner of the notebook. A new dialog box will appear.

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/2_share_button.png' width=500>

2. In the dialog box, make sure that '**Anyone on the Internet with this link can view**' option is selected and then click on the **Copy link** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/3_copy_link.png' width=500>

3. The link of the duplicate copy (named as **YYYY-MM-DD_StudentName_Project77**) of the notebook will get copied.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/4_copy_link_confirmation.png' width=500>

4. Go to your dashboard and click on the **My Projects** option.
   
   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/5_student_dashboard.png' width=800>

  <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/6_my_projects.png' width=800>

5. Click on the **View Project** button for the project you want to submit.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/7_view_project.png' width=800>

6. Click on the **Submit Project Here** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/8_submit_project.png' width=800>

7. Paste the link to the project file named as **YYYY-MM-DD_StudentName_Project77** in the URL box and then click on the **Submit** button.

   <img src='https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/images/project-share-images/9_enter_project_url.png' width=800>

---